#48 - Add basic controller for programatically controlling PieMenu

erecord opened the pull request:
# Summary
This PR adds a basic controller that exposes a `emitTapEvent()` method. When it's called, the pie menu will be opened at the mid point of the current widget.

# Changes
- Added PieMenuController
- Updated PieMenuCore to listen to the controller's ChangeNotifier and then synthetically tap the mid point of the current widget when the ChangeNotifier emits

Closes #48

rasitayaz reviewed (changes requested):
I had a multi-event controller in mind for future expandability. At the very least we should be able to emit both open and close events from the controller, just like we can go next and previous pages using PageController.

Also, we shouldn't be simulating tap events to open/close the menus; that would be lazy and innefficient. We can already open a menu using [`PieCanvasCore.attachMenu`](https://github.com/rasitayaz/flutter-pie-menu/blob/a92210c3884c35ab9219e2dd6f76c65b92e43c8e/lib/src/pie_canvas_core.dart#L494) method. Just need another method to close the menu by calling `PieCanvasCore._detachMenu` if the current menu on the canvas has the matching unique key.

erecord commented:
Hey @rasitayaz 
Thank you for your review and all this feedback. Now that I've made these changes I see how using `_notifier.canvas.attachMenu(...) directly is a better approach :)

I have updated the controller to be multi-event by using an enum. But let me know if you had another idea about implementing that.

Also, I have added `PieCanvasCore.closeMenu(Key)` to check for a key match and then call _detachMenu(). But calling _detachMenu(); doesn't seem to close the menu.

In my testing, _detachMenu(); is being called successfully when expected (passes the Key check), BUT it doesn't close the menu. Any ideas on this?

Let me know if I've misunderstood and interpreted anything incorrectly.

erecord commented:
I've got the controller.closeMenu() working now.

I needed to include ` _fadeController.reverse();` within `PieCanvasCore.closeMenu()`.

But, I'm worried that I'm not calling the following lines and that may cause issues? Do you think I should include these within the `PieCanvasCore.closeMenu()` method?
- _notifier.update(active: false);
- _notifyToggleListeners(active: false);

Feel free to review whenever you have some time. Thank you.

rasitayaz commented on lib/src/pie_menu_core.dart:
```diff
@@ -122,12 +127,20 @@ class _PieMenuCoreState extends State<PieMenuCore>
     _bounceController.dispose();
     _debounceTimer?.cancel();
     _bounceStopwatch.stop();
+    widget.controller?.dispose();
```
I think we shouldn't dispose the controller here since it's initialized outside by the user. We just need to remove the listener.

erecord replied:
Should be addressed now

rasitayaz commented:
Do you think can we transfer additional information like `Alignment` and `Offset` while sending the open signal? Maybe if we use data class instead of enum as the event we can include additional data like following:

```dart
abstract class PieMenuEvent {}

class PieMenuOpenEvent extends PieMenuEvent {
  Alignment menuAlignment;
  Offset menuDisplacement;
}
```

erecord commented:
Hey, thanks for the review. 

1. Good idea about not disposing the controller, nice spot!

2. > I've recently implemented a way to position the menu using an alignment and a displacement offset, if you are able to convert the event structure to empty classes I can handle the rest.
     - Yes, this Alignment idea is something I wanted to do in a follow-up PR but was just focusing on a small basic functionality for this PR - as a starting point to expand on later. For the 'empty classes' idea, could you provide a little more detail as I'm not 100% on what this would look like. Just a couple notes to help me on my way :)

rasitayaz commented:
I also added these alignment and displacement preferences to `PieTheme`, if you update your branch from dev you can utilize this:

https://github.com/rasitayaz/flutter-pie-menu/blob/2b2a761ce2a04bb73754ba84d3cb432657e0cf75/lib/src/pie_canvas_core.dart#L523-L538

We can pass `menuAlignment` and `menuDisplacement` as parameters to `attachMenu` and make null-aware assignments inside;

```dart
menuAlignment ??= _theme.menuAlignment;
menuDisplacement ??= _theme.menuDisplacement;
```

The rest will be handled by the canvas logic.

rasitayaz commented:
> I needed to include ` _fadeController.reverse();` within `PieCanvasCore.closeMenu()`. But am I just fading it without dismissing it?

Yeah that was a bad function call placement on my part, I now moved animation reversals to `_detachMenu` so you don't need to interact with `_fadeController` at all.

erecord commented on lib/src/pie_menu_controller.dart:
```diff
@@ -0,0 +1,22 @@
+  PieMenuController() : super(PieMenuCloseEvent());
+
+  void openMenu({
+    Alignment? menuAlignment,
```
I put these as nullable parameters because they could be available in PieTheme instead, and so are not required here. Assuming I've read your comments correctly and that's what you were after.

If you merge this branch in, feel free to change this around or delete it as you need.

erecord commented on lib/src/pie_menu_core.dart:
```diff
@@ -299,4 +303,38 @@ class _PieMenuCoreState extends State<PieMenuCore>
+
+  void _attachMenu(
+    bool rightClicked, {
+    PieMenuOpenEvent? menuOpenEvent,
```
I've added the `PieMenuOpenEvent? menuOpenEvent` here for you to hook into and use the event's alignment and displacement.

erecord commented:
I've now updated this branch so that it:
- Uses a class structure for events instead of an enum
- No longer interacts with _fadeController (thanks!)
- No longer has any custom positioning logic 
- Has Alignment and Displacement variables on the PieMenuOpenEvent
  - These are currently not used because you mentioned about setting this up and then you could handle the rest. 

Hopefully I've understood the change requests correctly. Let me know if anything needs changing though.

rasitayaz reviewed (approved):
looks great! I'll test it and release a new version.
